### PR TITLE
Hookmgr cleanup

### PR DESCRIPTION
**Pull request by Rochet2**
Rewriting hooks for new systems
Do not merge yet.

### CODE REVIEW

**Comment by Rochet2** [comment]
I think we still should return as soon as possible if there are no registered events.
Locking, pushing and popping all the values on each call to damage, kill, chat and so on will likely make a performance hit. And it is completely unnecessary to do so.

This system has not still been tested with calls within calls. Or at least I dont know of such.

File: HookMgr.cpp [hunk 1]
```diff
@@ -28,170 +28,273 @@ struct ScriptedAI;
...
+ */
+template<typename T>
+int Eluna::SetupStack(EventBind<T>* event_bindings, EntryBind<T>* entry_bindings, T event_id, uint32 entry, int number_of_arguments)
+{
+    // Ensure that if `entry_bindings` is not NULL, a valid entry is supplied.
+    ASSERT(!entry_bindings || (entry_bindings && entry > 0));
+    ASSERT(number_of_arguments == this->push_counter);
+    // Stack: [arguments]
+
+    Push((int)event_id);
```
Comment (Rochet2) [inline HookMgr.cpp]: Mostlikely no casting needed

**Comment by Rochet2** [comment]
I see object invalidation was removed. Was there a specific reason?
InvalidateObjects

**Comment by Rochet2** [comment]
I think instead of passing magical numbers and doing calculations, we could just do this:
```diff
void Eluna::OnReputationChange(Player* pPlayer, uint32 factionID, int32& standing, bool incremental)
{
    ELUNA_LOCK(this);
    Push(pPlayer);
    Push(factionID);
    Push(standing);
+   int standingIndex = lua_gettop(L);
    Push(incremental);
    int n = SetupStack(PlayerEventBindings, PLAYER_EVENT_ON_REPUTATION_CHANGE, 4);

    while (n > 0)
    {
        int r = CallOneFunction(n--, 4, 1);

        if (lua_isnumber(L, r))
        {
            standing = CHECKVAL<int32>(L, r);
            // Update the stack for subsequent calls.
+           ReplaceArgument(standing, standingIndex);
        }

        lua_pop(L, 1);
    }

    CleanUpStack(4);
}
```
ReplaceArgument would basically just be:
```c++
void Eluna::ReplaceArgument(T val, int index)
{
    Eluna::push(L, val);
    lua_replace(L, index);
}
```

**Comment by Rochet2** [comment]
Summing up chat here and skype:
1. Add InvalidateObjects
2. Reload is potentially unsafe if a packet is received. Make mutex global (static) to avoid deleting it in the middle of deleting eluna.
3. ReplaceArgument
4. Add RW locking with fast return if no functions to call

Last 2 not necessary, but would be nice.

**Comment by Patman64** [comment]
Should be ready now. Please compile/test.

**Pull request merged**